File: cmajor/parsing/Scanner.cpp
// =================================
// Copyright (c) 2019 Seppo Laakko
// Distributed under the MIT license
// =================================

#include <cmajor/parsing/Scanner.hpp>
#include <cmajor/parsing/Parser.hpp>
#include <cmajor/parsing/Exception.hpp>
#include <cmajor/parsing/Rule.hpp>
#include <algorithm>
#include <cctype>

namespace cmajor { namespace parsing {

std::u32string NarrowString(const char32_t* start, const char32_t* end)
{
    std::u32string s;
    s.reserve(end - start);
    while (start != end)
    {
        if (*start != '\r')
        {
            s.append(1, *start);
        }
        ++start;
    }
    return s;
}

bool countSourceLines = false;
int numParsedSourceLines = 0;

void SetCountSourceLines(bool count)
{
    countSourceLines = count;
}

int GetParsedSourceLines()
{
    return numParsedSourceLines;
}

Scanner::Scanner(const char32_t* start_, const char32_t* end_, const std::string& fileName_, int fileIndex_, Parser* skipper_):
    start(start_), end(end_), skipper(skipper_), skipping(false), tokenCounter(0), fileName(fileName_), span(fileIndex_),
    log(nullptr), atBeginningOfLine(true), parsingData(nullptr)
{
    if (countSourceLines)
    {
        numParsedSourceLines = 0;
    }
}

void Scanner::operator++()
{
    char32_t c = GetChar();
    if (countSourceLines)
    {
        if (!skipping && atBeginningOfLine && !std::isspace(c))
        {
            ++numParsedSourceLines;
            atBeginningOfLine = false;
        }
    }
    ++span;
    if (c == '\n')
    {
        atBeginningOfLine = true;
        span.IncLineNumber();
    }
}

void Scanner::Skip()
{
    if (tokenCounter == 0 && skipper)
    {
        Span save = span;
        BeginToken();
        skipping = true;
        ObjectStack stack;
        Match match = skipper->Parse(*this, stack, parsingData);
        skipping = false;
        EndToken();
        if (!match.Hit())
        {
            span = save;
        }
    }
}

int Scanner::LineEndIndex(int maxLineLength)
{
    int lineEndIndex = span.Start();
    int contentLength = static_cast<int>(end - start);
    int end = contentLength;
    if (maxLineLength != 0)
    {
        end = std::min(contentLength, span.Start() + maxLineLength);
    }
    while (lineEndIndex < end && (start[lineEndIndex] != '\r' && start[lineEndIndex] != '\n'))
    {
        ++lineEndIndex;
    }
    return lineEndIndex;
}

std::string Scanner::RestOfLine(int maxLineLength)
{
    std::string restOfLine(start + span.Start(), start + LineEndIndex(maxLineLength));
    return restOfLine;
}

} } // namespace cmajor::parsing

File: devenv/soul.parsing/Scanner.cs
/*========================================================================
    Copyright (c) 2011-2012 Seppo Laakko
    http://sourceforge.net/projects/soulparsing/
 
    Distributed under the GNU General Public License, version 3 (GPLv3).
    (See accompanying LICENSE.txt or http://www.gnu.org/licenses/gpl.html)

========================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace soul.parsing
{
    public class Scanner
    {
        public Scanner(string content, int file, string fileName, Parser skipper)
        {
            this.content = content;
            position.Index = 0;
            position.File = file;
            position.Line = 1;
            position.Column = 1;
            this.fileName = fileName;
            tokenCounter = 0;
            skipping = false;
            this.skipper = skipper;
        }
        public string Content
        {
            get { return content; }
        }
        public Position Position
        {
            get { return position; }
            set { position = value; }
        }
        public string FileName
        {
            get { return fileName; }
        }
        public int TokenCounter
        {
            get { return tokenCounter; }
        }
        public bool Skipping
        {
            get { return skipping; }
        }
        public Parser Skipper
        {
            get { return skipper; }
            set { skipper = value; }
        }
        public char Char
        {
            get { return content[position.Index]; }
        }
        public string Match(int startIndex)
        {
            return content.Substring(startIndex, position.Index - startIndex);
        }
        public bool AtEnd
        {
            get { return position.Index >= content.Length; }
        }
        public static Scanner operator ++(Scanner scanner)
        {
            char c = scanner.Char;
            ++scanner.position.Index;
            if (c == '\n')
            {
                ++scanner.position.Line;
                scanner.position.Column = 1;
            }
            else if (c != '\r')
            {
                ++scanner.position.Column;
            }
            return scanner;
        }
        public void BeginToken()
        {
            ++tokenCounter;
        }
        public void EndToken()
        {
            --tokenCounter;
        }
        public void Skip()
        {
            if (tokenCounter == 0 && skipper != null)
            {
                Position save = position;
                BeginToken();
                skipping = true;
                Match match = skipper.Parse(this, null);
                skipping = false;
                EndToken();
                if (!match.Hit)
                {
                    position = save;
                }
            }
        }
        public XmlLog Log
        {
            get { return log; }
            set { log = value; }
        }
        public int LineEndIndex
        {
            get
            {
                int lineEndIndex = position.Index;
                while (lineEndIndex < content.Length && (content[lineEndIndex] != '\r' && content[lineEndIndex] != '\n'))
                {
                    ++lineEndIndex;
                }
                return lineEndIndex;
            }
        }
        public string RestOfLine
        {
            get { return content.Substring(position.Index, LineEndIndex - position.Index);}
        }
        private string content;
        private Position position;
        private string fileName;
        private int tokenCounter;
        private bool skipping;
        private Parser skipper;
        private XmlLog log;
    }
}
